lib/core/src/sync/mod.rs

```diff
@@ -0,0 +1,214 @@
+pub(crate) mod client;
+pub(crate) mod model;
+
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use log::{debug, warn};
+use std::collections::HashSet;
+use tokio::sync::Mutex;
+use tokio_stream::StreamExt as _;
+
+use self::client::SyncerClient;
+use self::model::{
+    sync::{
+        ListChangesRequest, ListenChangesRequest, Record, SetRecordReply, SetRecordRequest,
+        SetRecordStatus,
+    },
+    DecryptedRecord, SyncData,
+};
+use crate::{model::Signer, persist::Persister, prelude::Direction, utils};
+
+const CURRENT_SCHEMA_VERSION: f32 = 0.01;
+
+pub(crate) struct SyncService {
+    connect_url: String,
+    persister: Arc<Persister>,
+    sent: Mutex<HashSet<i64>>,
+    signer: Arc<Box<dyn Signer>>,
+    client: Box<dyn SyncerClient>,
+}
+
+impl SyncService {
+    pub(crate) fn new(
+        connect_url: String,
+        persister: Arc<Persister>,
+        signer: Arc<Box<dyn Signer>>,
+        client: Box<dyn SyncerClient>,
+    ) -> Self {
+        Self {
+            connect_url,
+            persister,
+            signer,
+            client,
+            sent: Default::default(),
+        }
+    }
+
+    fn collect_records<'a>(
+        &self,
+        records: &'a [Record],
+    ) -> (Vec<DecryptedRecord>, Vec<&'a Record>) {
+        let mut failed_records = vec![];
+        let mut updatable_records = vec![];
+
+        for record in records {
+            // If it's a major version ahead, we skip
+            if record.schema_version.floor() > CURRENT_SCHEMA_VERSION.floor() {
+                failed_records.push(record);
+                continue;
+            }
+
+            let decrypted_record =
+                match DecryptedRecord::try_from_record(self.signer.clone(), record) {
+                    Ok(record) => record,
+                    Err(err) => {
+                        warn!("Could not decrypt record: {err:?}");
+                        continue;
+                    }
+                };
+
+            updatable_records.push(decrypted_record)
+        }
+
+        (updatable_records, failed_records)
+    }
+}
+
+impl SyncService {
+    pub(crate) async fn connect(&self) -> Result<()> {
+        self.client.connect(self.connect_url.clone()).await
+    }
+
+    pub(crate) async fn listen(self: Arc<Self>) -> Result<()> {
```

Comment:
NIT: from the client perspective it makes sense to have listen_changes but from the Syncer perhaps just start ? It does not only listen but also apply the changes and ensure we are up to sync.

Reply:
Since we need to call `apply_changes` which is a Syncer method, to avoid passing a reference to the Syncer on the Client we should keep the logic inside the Syncer.